File: lib/ui_loader.dart
import 'dart:html';

enum UIStates {
  init,
  data,
  prepare,
  group,
  sync,
  display,
  end
}

class UILoader {
  Element _element;
  Element _text;
  Map<UIStates, String> _states = {
    UIStates.init: 'Initializing',
    UIStates.data: 'Getting data from server',
    UIStates.prepare: 'Preparing locations',
    UIStates.group: 'Groupping branches by cities',
    UIStates.sync: 'Synchronizing OSM and NP branches',
    UIStates.display: 'Rendering',
    UIStates.end: 'Done'
  };

  UILoader(Element element) {
    _element = element;
    _text = _element.querySelector('.title');
  }

  setState(UIStates state, [String description]) {
    _element.classes.toggle('loading', state != UIStates.end);
    var text = _states[state];
    if (description != null)
      text += ' ($description)';
    _text.text = text;
    // Force render wait.
    return window.animationFrame;
  }
}

File: lib/ui_loader.ts
class UILoader {
  _element : Element;
  _text : Element;
  _states : { [email: string]: string; } = {
    'init': 'Initializing',
    'data': 'Getting data from server',
    'prepare': 'Preparing locations',
    'group': 'Groupping branches by cities',
    'sync': 'Synchronizing OSM and NP branches',
    'display': 'Rendering',
    'end': 'Done'
  };

  constructor(element: Element) {
    this._element = element;
    this._text = this._element.querySelector('.title');
  }

  async setState(state: string, description?: string) {
    this._element.classList.toggle('loading', state != 'end');
    var text = this._states[state];
    if (description != null)
      text += ' ($description)';
    this._text.textContent = text;
    return new Promise<void>(resolve => {
      window.requestAnimationFrame(resolve);
    });
  }
}